org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/GradleClasspathContainerInitializer.java

```diff
@@ -171,6 +172,12 @@ public IClasspathEntry apply(OmniEclipseSourceDirectory directory) {
                 // @formatter:on
             }
         }).toList();
+
+        return removeDuplicatesFromSourceFolders(sourceFolders);
+    }
+
+    private List<IClasspathEntry> removeDuplicatesFromSourceFolders(List<IClasspathEntry> sourceFolders) {
+        return ImmutableSet.copyOf(sourceFolders).asList();
```

Comment:
I've answered this question in our conversion, but I'll post the reply here for posterity:

ImmutableSet.copyOf() keeps the order - http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/ImmutableSet.html